SQLite: Foreign Key Support

kamipo opened the pull request:
https://www.sqlite.org/foreignkeys.html

sgrif commented on activerecord/test/schema/schema.rb:
```diff
@@ -992,16 +992,19 @@
-    create_table :fk_test_has_fk, force: true do |t|
-      t.integer :fk_id, null: false
-    end
+  if supports_foreign_keys_in_create?
```
We shouldn't be pushing caring about this onto our users. If `create_table` is called with foreign keys on SQLite, we should do a separate query automatically

kamipo replied:
If `create_table` is called with foreign keys, generates in single query. ref #20009.

sgrif replied:
I know. You're missing my point. It needs to not do that on SQLite. This code should be capable of running on both backends.

kamipo replied:
Sorry I don't understand. AFAIK this code block works on mysql2, postgresql, and sqlite3 adapters.

```ruby
  if supports_foreign_keys_in_create?
    drop_table :fk_test_has_fk, if_exists: true

    create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
    end

    create_table :fk_test_has_fk do |t|
      t.integer :fk_id, null: false
      t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
    end
  end
```

MySQL:

```sql
DROP TABLE IF EXISTS `fk_test_has_fk`
DROP TABLE `fk_test_has_pk`
CREATE TABLE `fk_test_has_pk` (`pk_id` int AUTO_INCREMENT PRIMARY KEY) ENGINE=InnoDB
CREATE TABLE `fk_test_has_fk` (`id` int AUTO_INCREMENT PRIMARY KEY, `fk_id` int NOT NULL, CONSTRAINT `fk_name`
FOREIGN KEY (`fk_id`)
  REFERENCES `fk_test_has_pk` (`pk_id`)
) ENGINE=InnoDB
```

PostgreSQL:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" serial primary key)
CREATE TABLE "fk_test_has_fk" ("id" serial primary key, "fk_id" integer NOT NULL, CONSTRAINT "fk_name"
FOREIGN KEY ("fk_id")
  REFERENCES "fk_test_has_pk" ("pk_id")
)
```

SQLite3:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL)
CREATE TABLE "fk_test_has_fk" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "fk_id" integer NOT NULL, CONSTRAINT "fk_name"
FOREIGN KEY ("fk_id")
  REFERENCES "fk_test_has_pk" ("pk_id")
)
```

sgrif replied:
What I'm saying is this should work without the if statement. Pn versions of SQLite that do not support foreign keys in create, or any other adapter that doesn't support this as a single query, it should automatically perform 2 queries.

kamipo commented on activerecord/test/schema/schema.rb:
```diff
@@ -992,16 +992,17 @@
+  create_table :fk_test_has_fk do |t|
+    t.integer :fk_id, null: false
+    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
+  end
```
@sgrif Thank you for explanation, I understand. Actually this works without `if supports_foreign_keys_in_create?`. I removed the if statement.

sgrif replied:
Tested on an older version of SQLite?

kamipo replied:
```ruby
def (ActiveRecord::Base.connection).sqlite_version
  @sqlite_version ||= ActiveRecord::ConnectionAdapters::SQLite3Adapter::Version.new('3.6.16')
end

ActiveRecord::Schema.define do
  drop_table :fk_test_has_fk, if_exists: true

  create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
  end

  create_table :fk_test_has_fk do |t|
    t.integer :fk_id, null: false
    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
  end
end
```

Result:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL)
CREATE TABLE "fk_test_has_fk" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "fk_id" integer NOT NULL)
```

https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L52-L54
https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L910

Even Ubuntu 10.04 bundles sqlite3 3.6.22. To test an actual older version (< 3.6.19) is difficult for me...
https://launchpad.net/ubuntu/+source/sqlite3/3.6.22-1

kamipo commented on activerecord/test/schema/schema.rb:
```diff
@@ -992,19 +992,18 @@
-    add_foreign_key :lessons_students, :students
+  create_table :fk_test_has_fk do |t|
+    t.integer :fk_id, null: false
+    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
```
> Tested on an older version of SQLite?

```ruby
def (ActiveRecord::Base.connection).sqlite_version
  @sqlite_version ||= ActiveRecord::ConnectionAdapters::SQLite3Adapter::Version.new('3.6.16')
end

ActiveRecord::Schema.define do
  drop_table :fk_test_has_fk, if_exists: true

  create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
  end

  create_table :fk_test_has_fk do |t|
    t.integer :fk_id, null: false
    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
  end
end
```

Result:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL)
CREATE TABLE "fk_test_has_fk" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "fk_id" integer NOT NULL)
```

https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L52-L54
https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L910

Even Ubuntu 10.04 bundles sqlite3 3.6.22. To test an actual older version (< 3.6.19) is difficult for me...
https://launchpad.net/ubuntu/+source/sqlite3/3.6.22-1

kamipo committed:
SQLite: Foreign Key Support

https://www.sqlite.org/foreignkeys.html

kamipo committed:
Should work foreign key in test schema without `if supports_foreign_keys?` statement

If an adapter does not support foreign key feature, should be noop.

https://github.com/rails/rails/blob/v5.0.0.rc1/activerecord/test/cases/migration/foreign_key_test.rb#L288-L294
https://github.com/rails/rails/blob/v5.0.0.rc1/activerecord/test/cases/migration/references_foreign_key_test.rb#L208-L214

kamipo commented:
Since you has merged #27701, and #24986 has assigned to you, can you also review the PR for foreign key creation? r? @pixeltrix

pixeltrix commented:
@kamipo doesn't this include the changes in #24986

pixeltrix commented:
@kamipo thanks! 👍
